compiler/vm/vm.nim

```diff
@@ -2852,19 +2823,51 @@ proc rawExecute(c: var TCtx, pc: var int, tos: var StackFrameIndex): YieldReason
       inc pc
       let srctyp = c.rtti[c.code[pc].regBx - wordExcess]
 
-      # No need to check the handle of regs[ra] since it's not accessed
+      checkHandle(regs[ra])
       checkHandle(regs[rb])
 
       # TODO: `opConv` needs to accept `VmTypeInfo` directly
       let dt = (desttyp.nimType, desttyp.internal)
       let st = (srctyp.nimType, srctyp.internal)
       if opConv(c, regs[ra], regs[rb], dt, st):
         raiseVmError(VmEvent(
-          kind: vmEvtIllegalConv,
-          msg: errIllegalConvFromXtoY % [
-            typeToString(srctyp.nimType),
-            typeToString(desttyp.nimType)
-          ]))
+          kind: vmEvtIllegalConvFromXToY,
+          typeMismatch: VmTypeMismatch(
+            actualType: srctyp.nimType,
+            formalType: desttyp.nimType)))
+    of opcObjConv:
+      # a = conv(b)
+      ensureKind(rkHandle)
+      let src = regs[instr.regB].handle
+      inc pc
+      let desttyp = c.types[c.code[pc].regBx - wordExcess]
+
+      assert src.typ.kind == desttyp.kind
+      # perform a conversion check, if necessary
+      case desttyp.kind
+      of akRef:
+        # a ref up- or down conversion
+        assert src.typ.kind == akRef # vmgen issue
+        checkHandle(regs[instr.regB])
```

Comment:
A `checkHandle` that takes the handle itself already exists, but I didn't use it for efficiency reasons. The overload taking a register doesn't perform access checks for owning handles (i.e. `rkLocation` -- they're always valid), while the one taking a handle always does.

Further down the road, both the `rkLocation` and all the `checkHandle` calls will go away (the latter will be replaced with a dedicated operation, so that the code generator can decide on the placement).

Reply:
Sorry, I should have read the implementation a bit more closely, they looked equivalent except for the handle fetch. Thanks for the explanation.